### ISSUE TITLE
Problems regarding the timingPattern when calling diva2gtfs.pl

### ISSUE THREAD

**Issue by nikolauskrismer** [stkdiretto/diva2gtfs#3]
I do get some errors when calling diva2gtfs.
The one that seems to be repeated all the time (which makes it impossible for me to see if there are actually more than one error) is:

```
Use of uninitialized value in string ne at ./diva2gtfs.pl line 268, <FILE> chunk 1037.
Use of uninitialized value in string ne at ./diva2gtfs.pl line 268, <FILE> chunk 1037.
Use of uninitialized value in string ne at ./diva2gtfs.pl line 268, <FILE> chunk 1037.
Use of uninitialized value in addition (+) at ./diva2gtfs.pl line 271, <FILE> chunk 1037.
Use of uninitialized value in concatenation (.) or string at ./diva2gtfs.pl line 301, <FILE> chunk 1037.
```

Is there some freely available test-data where I can compare my file (from uvz/lierg.kbez) with the expected format? To me the data looks good (the timePattern is there and seems valid to me, although I am not sure how it should be defined in the file's header)...

Maybe we can figure this out together, so I can send another pull request :-)

**Comment by nikolauskrismer** [comment]
After some debugging my first guess is that the timing patterns are not found by using the regexp provided in the script.

**Comment by stkdiretto** [comment]
Unfortunately, I do not have test data anymore. Maybe(!) changing line 177

`elsif ($line =~ /FT(?<ftid>[HR][0-9]{5}).{2}(?<pattern>.*)[ ,N].* .*/) {`

to 

`elsif ($line =~ /FT(?<ftid>[HR][0-9]{5}).{2}(?<pattern>.*)[ ,N].*/) {`

might help. I have no idea how the whitespace thing came into this RegExp, but it seems to help with the test data within the thesis (which is the only one I have left, as per the NDA). Alternatively,

`elsif ($line =~ /FT(?<ftid>[HR][0-9]{5}).{2}(?<pattern>.*)[ ,N].* ?.*/) {`

might work.

**Comment by nikolauskrismer** [comment]
Changing the line did help, and timing patterns gets parsed now.
However, there is still something wrong, because there are more stops than timingpatterns (according to your code the number should match, isn't it?)

I guess the problem is related to stop pattern matching and not timingpattern matching, but I am not sure (however, stops look more suspicious, since my last stop only contains two digits, while your pattern suggests that all stops should have four digits).

Maybe this is also related to stop platform handling (there is an open TODO task in your code)

**Comment by stkdiretto** [comment]
The stop pattern should say right at the beginning how many stops it is going to define. 

E.g., `STH012` should be followed by 12 stops.

Do you have a printed timetable on hand, to compare the data you have to what it should looke like in the wild?

**Comment by nikolauskrismer** [comment]
Yes... that was the problem... my steps do contain 5 digits instead of 4.
I changed to code to dynamically calculate the length and now most of the perl errors have gone away.
However, at the moment I still get a lot of "Trip handling failed" messages.... I'll have a look at that after lunch or in the evening :+1:

**Comment by nikolauskrismer** [comment]
The problem might be easier to debug than I thought:
the failing trips do not start with FAH0, but with FAHL, FAHM, FAHN (or FARL, FARM and FARN for the opposite direction)...
...I have no idea why there is a L/M/N -> is this the serviceId and/or the journeykey that contains a char (and therefor is not recognized by the pattern that only takes care of numbers)??

**Comment by stkdiretto** [comment]
Oh, yes, that might be it – see [Section 5.4.5 (Page 65)](http://dbis.eprints.uni-ulm.de/1054/1/Kaufmann2014.pdf); within the data I had, only the digits 0, 2 and 3 were used as Day of Week identifiers (“service id”). You might encounter this problem also when headsign texts are looked up (strings starting with `EE`).

**Comment by nikolauskrismer** [comment]
Problems should be fixed by PR.

The tests of the changes were successful, but I am still not done with creating a GTFS file from a DIVA file (transfers2gtfs.pl seems to take quite some time - at least for my tests data)

**Comment by nikolauskrismer** [comment]
The problem is only partially fixed.
The regexp has been successfully fixed with the changes included in PR #4.

However, I now have trips that reference a service_id that does not exist.
For me the service_ids not defined are: 1,4,8,9,A,G,H,L,M and N

Are these hard-coded service definitions? If so... where can I find the corresponding hard-coded calendars?
If not... from where can I read the calendars?

The service2gtfs.pl script creates two-letter service ids and the service ids 0,2 and 3, but none of the mentioned above ;-(

**Comment by stkdiretto** [comment]
I wouldn't know about any of these. So, to make things clear,
- you encounter journey definitions starting with `FA[H,R][1,4,8,9,A,G,H,L,M,N]`
- those service ids are not defined within `ServiceRestriction`
- there is no other source table within your original DIVA feed that gives any clues about those service ids?

**Comment by nikolauskrismer** [comment]
Yes... thats true.

My journey definition looks like this (for service id  'N'): FAHN144400000000314442000033
"L" is not included in the ServiceRestriction table... sry :-(
I did not find anything about "L" (when grepping over the files). Looking at the data by hand does not reveal useful information (maybe I overlooked it... hard to say when looking by hand).

**Comment by nikolauskrismer** [comment]
Hmm... might have found a clue in the ServiceRestriction table (DIVA database).
Entries with code statring with either "T" or "U" are not used directly... however, there is T1, T4, T8, T9, TA (also TC, which seems to be unused) and there is UG, UH, UI. So maybe skipping the first character for these service restrictions is what needs to be done. I am not sure if this is hard-coded to T and U are maybe applies to all upper-case letters (all my other service restrictions that are used directly use lower-case).

By the way:
Is there a reason why everything is added to calendar_dates and not do calendar? All my entries in the gtfs calendar table contain only 0 for every day (except for the hard-coded service_ids 0,2 and 3).
Shouldn't we try to handle as much in calendar table (and not by calendar_dates)?

**Comment by nikolauskrismer** [comment]
Fixed by #8 (by the change that ServiceRestrictions with kenn 'I' are used now with skipping the first char)
